src/libraries/Common/src/Roslyn/MetadataHelpers.cs

```diff
@@ -0,0 +1,1034 @@
+
+                        case ']':
+                            if (arrayRanksBuilder == null)
+                            {
+                                arrayRanksBuilder = ArrayBuilder<int>.GetInstance();
+                            }
+
+                            arrayRanksBuilder.Add(rank == 1 && !isMultiDimensionalIfRankOne ? 0 : rank);
+                            Advance();
+                            return;
+
+                        case '*':
+                            if (rank != 1)
+                            {
+                                goto default;
+                            }
+
+                            Advance();
+                            if (Current != ']')
+                            {
+                                // Error case, process as regular characters
+                                typeNameBuilder.Append(_input.Substring(start, _offset - start));
+                                return;
+                            }
+
+                            isMultiDimensionalIfRankOne = true;
+                            break;
+
+                        default:
+                            // Error case, process as regular characters
+                            Advance();
+                            typeNameBuilder.Append(_input.Substring(start, _offset - start));
+                            return;
+                    }
+                }
+
+                // Error case, process as regular characters
+                typeNameBuilder.Append(_input.Substring(start, _offset - start));
+            }
+        }
+
+        private static readonly string[] s_aritySuffixesOneToNine = { "`1", "`2", "`3", "`4", "`5", "`6", "`7", "`8", "`9" };
+
+        internal static string GetAritySuffix(int arity)
+        {
+            Debug.Assert(arity > 0);
+            return (arity <= 9) ? s_aritySuffixesOneToNine[arity - 1] : string.Concat(GenericTypeNameManglingString, arity.ToString(CultureInfo.InvariantCulture));
+        }
+
+        internal static string ComposeAritySuffixedMetadataName(string name, int arity)
+        {
+            return arity == 0 ? name : name + GetAritySuffix(arity);
+        }
+
+#endif
+
+        internal static int InferTypeArityFromMetadataName(string emittedTypeName)
+        {
+            return InferTypeArityFromMetadataName(emittedTypeName, out _);
+        }
+
+        private static short InferTypeArityFromMetadataName(string emittedTypeName, out int suffixStartsAt)
+        {
+            Debug.Assert(emittedTypeName != null, "NULL actual name unexpected!!!");
+            int emittedTypeNameLength = emittedTypeName.Length;
+
+            int indexOfManglingChar;
+            for (indexOfManglingChar = emittedTypeNameLength; indexOfManglingChar >= 1; indexOfManglingChar--)
+            {
+                if (emittedTypeName[indexOfManglingChar - 1] == GenericTypeNameManglingChar)
+                {
+                    break;
+                }
+            }
```

Comment:
Did you consider using an API like LastIndexOf rather than open-coding this?

Reply:
so this is just a straight port of Roslyn: https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs#L484

so i def didn't want to touch this and screw anything up here *at all* :)

Reply:
Yup, wasn't suggesting it being changed here.  I'm more wondering why Roslyn isn't relying on such helpers, if there's an issue with them or if it just didn't happen, etc.

Reply:
it's a good question.  my recollection is that some of this are fairly direct ports of hte old native code.  So it may just have been super expedient to copy virtually unchanged and then have it not be touched.  A case of "if it ain't broke".  Given the nicer runtime APIs (and all the recent improvements around strings/spans as well), this likely could get an entire makeover.  So it would more a question of if Compiler team would see value in that, our would prefer to not touch and put attention on other more valuable stuff :)